core/src/main/java/io/grpc/Metadata.java

```diff
@@ -132,87 +135,145 @@ public Integer parseAsciiString(String serialized) {
-        break;
-      }
-      String name = new String(binaryValues[i], US_ASCII);
-      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[i + 1]));
-    }
+  Metadata(int usedNames, byte[]... binaryValues) {
+    // TODO(carl-mastrangelo): maybe remove this, unless
+    // https://bugs.openjdk.java.net/browse/JDK-6445664 is fixed.
+    assert (binaryValues.length & 1) == 0 : "Odd number of key-value pairs " + binaryValues.length;
+    size = usedNames;
+    namesAndValues = binaryValues;
+    for (int i = 0; i < usedNames * 2; i++) {
+      checkNotNull(namesAndValues[i], "val was null %s/%s", i, namesAndValues.length);
+    }
+    typedValues = size != 0 ? new Object[size] : null;
+  }
+
+  private byte[][] namesAndValues;
+  private Object[] typedValues;
+  // The unscaled number of headers present.
+  private int size;
+
+  private byte[] name(int i) {
+    return namesAndValues[i * 2];
+  }
+
+  private void name(int i, byte[] name) {
+    namesAndValues[i * 2] = name;
+  }
+
+  private byte[] value(int i) {
+    return namesAndValues[i * 2 + 1];
+  }
+
+  private void value(int i, byte[] value) {
+    namesAndValues[i * 2 + 1] = value;
+  }
+
+  private int cap() {
+    return namesAndValues != null ? namesAndValues.length : 0;
+  }
+
+  // The scaled version of size.
+  private int len() {
+    return size * 2;
   }
 
   /**
    * Constructor called by the application layer when it wants to send metadata.
    */
   public Metadata() {}
 
-  private void storeAdd(String name, MetadataEntry value) {
-    List<MetadataEntry> values = store.get(name);
-    if (values == null) {
-      // We expect there to be usually unique header values, so prefer smaller arrays.
-      values = new ArrayList<MetadataEntry>(1);
-      store.put(name, values);
-    }
-    storeCount++;
-    values.add(value);
-  }
-
   /**
    * Returns the total number of key-value headers in this metadata, including duplicates.
    */
   @Internal
   public int headerCount() {
-    return storeCount;
+    return size;
   }
 
   /**
    * Returns true if a value is defined for the given key.
    */
   public boolean containsKey(Key<?> key) {
-    return store.containsKey(key.name());
+    for (int i = 0; i < size; i++) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        return true;
+      }
+    }
+    return false;
   }
 
   /**
    * Returns the last metadata entry added with the name 'name' parsed as T.
    * @return the parsed metadata entry or null if there are none.
    */
   public <T> T get(Key<T> key) {
-    List<MetadataEntry> values = store.get(key.name());
-    if (values == null) {
-      return null;
+    for (int i = size - 1; i >= 0; i--) {
+      if (Arrays.equals(key.asciiName(), name(i))) {
+        if (typedValues[i] == null) {
+          typedValues[i] = key.parseBytes(value(i));
+        }
+        @SuppressWarnings("unchecked") // take THAT typesafety!
+        T ret = (T) typedValues[i];
```

Comment:
This isn't correct, since the keys may use different marshallers but have the same name. The previous code stored the associated key so that it could determine that the value was still correct.